cloudflare/resource_cloudflare_access_identity_provider.go

```diff
@@ -0,0 +1,329 @@
+
+	return resourceCloudflareAccessIdentityProviderRead(d, meta)
+}
+
+func resourceCloudflareAccessIdentityProviderUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	accountID := d.Get("account_id").(string)
+
+	IDPConfig, conversionErr := convertSchemaToStruct(d)
+	if conversionErr != nil {
+		return fmt.Errorf("failed to convert schema into struct: %s", conversionErr)
+	}
+
+	log.Printf("[DEBUG] updatedConfig: %+v", IDPConfig)
+	updatedAccessIdentityProvider := cloudflare.AccessIdentityProvider{
+		Name:   d.Get("name").(string),
+		Type:   d.Get("type").(string),
+		Config: IDPConfig,
+	}
+
+	log.Printf("[DEBUG] Updating Cloudflare Access Identity Provider from struct: %+v", updatedAccessIdentityProvider)
+
+	accessIdentityProvider, err := client.UpdateAccessIdentityProvider(accountID, d.Id(), updatedAccessIdentityProvider)
+	if err != nil {
+		return fmt.Errorf("error updating Access Identity Provider for ID %q: %s", d.Id(), err)
+	}
+
+	if accessIdentityProvider.ID == "" {
+		return fmt.Errorf("failed to find Access Identity Provider ID in update response; resource was empty")
+	}
+
+	return resourceCloudflareAccessIdentityProviderRead(d, meta)
+}
+
+func resourceCloudflareAccessIdentityProviderDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*cloudflare.API)
+	accountID := d.Get("account_id").(string)
+
+	log.Printf("[DEBUG] Deleting Cloudflare Access Identity Provider using ID: %s", d.Id())
+
+	_, err := client.DeleteAccessIdentityProvider(accountID, d.Id())
+	if err != nil {
+		return fmt.Errorf("error deleting Access Policy for ID %q: %s", d.Id(), err)
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func resourceCloudflareAccessIdentityProviderImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	attributes := strings.SplitN(d.Id(), "/", 2)
+
+	if len(attributes) != 2 {
+		return nil, fmt.Errorf("invalid id (\"%s\") specified, should be in format \"accountID/accessIdentityProviderID\"", d.Id())
+	}
+
+	accountID, accessIdentityProviderID := attributes[0], attributes[1]
+
+	log.Printf("[DEBUG] Importing Cloudflare Access Identity Provider: accountID=%s accessIdentityProviderID=%s", accountID, accessIdentityProviderID)
+
+	d.Set("account_id", accountID)
+	d.SetId(accessIdentityProviderID)
+
+	resourceCloudflareAccessIdentityProviderRead(d, meta)
+
+	return []*schema.ResourceData{d}, nil
+}
+
+func convertSchemaToStruct(d *schema.ResourceData) (cloudflare.AccessIdentityProviderConfiguration, error) {
+	IDPConfig := cloudflare.AccessIdentityProviderConfiguration{}
+
+	if _, ok := d.GetOk("config"); ok {
+		if _, ok := d.GetOk("config.0.attributes"); ok {
+			attrData := make([]string, d.Get("config.0.attributes.#").(int))
+			for id := range attrData {
+				attrData[id] = d.Get(fmt.Sprintf("config.0.attributes.%d", id)).(string)
+			}
+			IDPConfig.Attributes = attrData
+		}
+
+		IDPConfig.AppsDomain = d.Get("config.0.apps_domain").(string)
```

Comment:
Yes, this is a little verbose but I don't mind it as it's explicit and it doesn't require confusing reflection to work out what each field should be.